tools/pre-push

```diff
@@ -5,21 +5,21 @@
 #   cd .git/hooks # from repo root
 #   ln -s ../../tools/pre-push .
 
-if ! git diff HEAD --quiet; then
```

Comment:
Don't take away the functionality to be able to push with a dirty staging area. I value this functionality over seeing the exact style changes required; on a local environment, it's easy enough to just run `make style` and make styling issues go away.

Reply:
I get the sense this is your primary objection to this pull-request. If the other changes in this P.R. to broaden `make lint` to be comprehensive are merged then I have no objection to keeping the current pre-push hook behavior; although, I admit I don't understand why you want to allow pushing with a dirty staging area. But that's no doubt because you and I employ different Git workflows. If my staging area is dirty, but I still want to push, I expect to manually perform a `git stash` and would be surprised by the pre-push hook doing that automatically. But that's just me. 😄